### PR TITLE
Add class min threshold value property to model inference node

### PR DESCRIPTION
**Pull request by SIGLOST**

### CODE REVIEW

File: pipeline/src/node_properties/model_inference_properties.rs [hunk 1]
```diff
@@ -17,6 +17,9 @@ pub struct ModelInferenceProperties {
 
     #[serde(flatten)]
     pub runtime: Option<ModelInferenceRuntime>,
+
+    /// Optional vector containing the min inference threshold value for each class.
+    pub classes_thresholds: Option<Vec<PerClassThreshold>>,
```
Comment (DmitrySamoylov) [inline pipeline/src/node_properties/model_inference_properties.rs:22]: Can it be just a hashmap `class_id -> min_threshold`?

Comment (SIGLOST) [reply pipeline/src/node_properties/model_inference_properties.rs:22]: While digging in that question I found some rust profiling results that shows faster lookup using a vector than a hashmap in the case we have a small number of elements (less than 15 - which actually is the case for the majority of models classes, with the exception of yolo) 

Anyway I don't have strong opinion on that :), in the end of the day it should not be big performance difference as we want to use it mostly for full iteration and not for lookup

Comment (SIGLOST) [reply pipeline/src/node_properties/model_inference_properties.rs:22]: The idea of optional field was handle those cases:
- if there are no `classes_thresholds` defined on console's inference node, then just load the default configuration
- otherwise if there's `classes_thresholds` create user custom config 

ofc we can just default to empty vector and later add additional check in other components checking if it's empty or not

**Review by DmitrySamoylov — APPROVED** [review]

File: pipeline/src/node_properties/model_inference_properties.rs [hunk 2]
```diff
@@ -39,3 +43,13 @@ pub struct ModelInferenceRuntime {
...
     pub infer_node_unique_ids: Option<BTreeMap<String, i32>>,
 }
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub struct PerClassThreshold {
+    /// Use a negative class_id value (i.e: -1) to define a global threshold for all the classes.
+    ///
+    /// To define individual class threshold, we accept integers starting from 0 using the same
+    /// order as the model class labels.
+    pub class_id: i32,
```
Comment (devarshi) [inline pipeline/src/node_properties/model_inference_properties.rs:53]: @SIGLOST Thinking out loud - wondering if it would be more scalable if we used class labels instead of id's here. Here's the rationale 
- for nvinfer, the thresholds are specified using class id's, but some other inference element (ex https://github.com/openvinotoolkit/dlstreamer_gst/wiki/gvainference ) may require differently ?
- should console / API have to know how to map labels to class id's ? if we specify this using class labels, then this mapping can live in lumeod and api/console can deal with the high level specification using labels.

Comment (SIGLOST) [reply pipeline/src/node_properties/model_inference_properties.rs:53]: So if the inference node in the future might accept both nvinfer (that takes "class_id") and other inference element (that takes "class_label") I think the more flexible option for the API would be expand this struct to have both "class_id" and "class_label".

The "class_id" is an integer starting from 0 that uses the same order as the classes are defined so should not be difficult to populate on the root element that retrieves this info. 

That way console and API users would easily fill the required fields, depending on the circumstances. and backend components would not be required to read and parse additional files (maybe multiple times).

If that means a lot of work or we have some blocker, ofc we can switch here to just "class_label" and propagate the mapping for the backend components

Comment (devarshi) [reply pipeline/src/node_properties/model_inference_properties.rs:53]: > The "class_id" is an integer starting from 0 that uses the same order as the classes are defined so should not be difficult to populate on the root element that retrieves this info.

This is an implicit assumption specific to nvidia's implementation - so IMHO it is likely a wise thing to leave it out. From a user standpoint, and from a cloud standpoint, labels are the ones that really matter.

Comment (SIGLOST) [reply pipeline/src/node_properties/model_inference_properties.rs:53]: Yes that's true the end-user (at console level) just cares about the "labels", but the usage of "class_id" (or an integer class) is widely used and attached to the inference engines layer, as with them is always possible to make a 1:1 relation with the neural network output. This approach is used mainly because depending on the implementation you can map the int output to different needs or languages (english, chinese, etc). In fact nvinfer takes a smart approach as it opens the possibility of each user customize the classes names they want to use, without causing any impact on the configuration.

For example in tensorflow you have some utils to later associate of the integer classes with labels (basically a list of strings following the same order): 
https://www.tensorflow.org/datasets/api_docs/python/tfds/features/ClassLabel

Comment (SIGLOST) [reply pipeline/src/node_properties/model_inference_properties.rs:53]: on Openvino:
https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/samples/object_detection_sample_ssd/main.cpp#L323

I didn't found any high-level config on this framework related with per-class config on the gva component
